litex/build/gowin/gowin.py

```diff
@@ -0,0 +1,113 @@
+#
+# This file is part of LiteX.
+#
+# Copyright (c) 2020 Pepijn de Vos <pepijndevos@gmail.com>
+# Copyright (c) 2015-2018 Florent Kermarrec <florent@enjoy-digital.fr>
+# SPDX-License-Identifier: BSD-2-Clause
+
+import os
+import subprocess
+
+from migen.fhdl.structure import _Fragment
+
+from litex.build.generic_platform import Pins, IOStandard, Misc
+from litex.build import tools
+
+def _build_cst(named_sc, named_pc):
+    lines = []
+
+    flat_sc = []
+    for name, pins, other, resource in named_sc:
+        if len(pins) > 1:
+            for i, p in enumerate(pins):
+                flat_sc.append((f"{name}[{i}]", p, other))
+        else:
+            flat_sc.append((name, pins[0], other))
+
+    for name, pin, other in flat_sc:
+        lines.append(f"IO_LOC \"{name}\" {pin};")
+    
+        for c in other:
+            if isinstance(c, IOStandard):
+                lines.append(f"IO_PORT \"{name}\" IO_TYPE={c.name};")
+            elif isinstance(c, Misc):
+                lines.append(f"IO_PORT \"{name}\" {c.misc};")
+
+    if named_pc:
+        lines.extend(named_pc)
+
+    cst = "\n".join(lines)
+    with open("top.cst", "w") as f:
+        f.write(cst)
+
+def _build_script(name, partnumber, files, options):
+    lines = [
+        f"set_device -name {name} {partnumber}",
+        "add_file top.cst",
+    ]
+
+    for f, typ, lib in files:
+        lines.append(f"add_file {f}")
+
+    for opt, val in options.items():
+        lines.append(f"set_option -{opt} {val}")
+
+    lines.append("run all")
+
+    tcl = "\n".join(lines)
+    with open("run.tcl", "w") as f:
+        f.write(tcl)
+
+class GowinToolchain():
+    def __init__(self):
+        self.options = {}
+
+    def attr_translate(*args):
```

Comment:
You can copy this: https://github.com/enjoy-digital/litex/blob/master/litex/build/microsemi/libero_soc.py#L219-L230

Reply:
I copied the code for `get_verilog` from somewhere that had it, but saw that another target did not have it at all. I'm not clear what this function does, and what the correct implementation would be for Gowin. I actually thought I deleted the entire function. So yea I can copy those lines, but I'd like to understand why it's needed and have some code that test the functionality.

Reply:
This is used to translate attributes that we add to internal signals in the Migen/LiteX design to verilog attributes (that are different depending the toolchain). The `keep` attribute for example is used to specify that the signal should not be simplified during synthesis and we are generally using it to specify timing constraints. Most of these attributes are only used with the Xilinx toolchains and this should be simplified/refactored on the other toolchains, but i just suggest you copy the same structure to ease refactoring later.

Reply:
Alright, added dummy translations.